### PR TITLE
Refactor to use listeners in deployment reporter

### PR DESCRIPTION
**Pull request by agusaldasoro**
This PR fixes the issue in https://github.com/decentraland/catalyst/issues/203

### CODE REVIEW

File: content/src/service/reporters/SQSDeploymentReporter.ts [hunk 1]
```diff
@@ -2,27 +2,46 @@ import log4js from "log4js";
...
+                region: "us-east-1",
+                credentials: new Credentials(sqsKey, sqsSecret)
+            })
+            service.listenToDeployments((deployment) => this.onDeployment(deployment))
+        }
+
     }
 
+
+    private async onDeployment(deploymentEvent: DeploymentEvent): Promise<void> {
```
Comment (agusaldasoro) [inline content/src/service/reporters/SQSDeploymentReporter.ts:38]: @nchamo any idea on how to avoid having this method duplicated on both classes?

File: content/src/service/reporters/SQSDeploymentReporter.ts [hunk 2]
```diff
@@ -2,27 +2,46 @@ import log4js from "log4js";
...
     constructor(
-        accessKeyId: string,
-        secretAccessKey,
-        private readonly queueURL: string,
-        private readonly callback?: (error?:string, messageId?:string) => void) {
-
-        this.sqsClient = new SQS({
-            region: "us-east-1",
-            credentials: new Credentials(accessKeyId, secretAccessKey)})
+        env: Environment,
```
Comment (nchamo) [inline content/src/service/reporters/SQSDeploymentReporter.ts:19]: Let's not take the `Environment` here, and just read everything on the factory. It's easier for mocking purposes. We can pass the different variables as we did before, but they would be of type `string | undefined` now.

File: content/src/service/reporters/SQSDeploymentReporter.ts (hunk 1 above)
Comment (nchamo) [reply content/src/service/reporters/SQSDeploymentReporter.ts:38]: On line 32, you could just write:
```
service.listenToDeployments((deployment) => this. reportDeployment(deployment.entity, Authenticator.ownerAddress(deployment.auditInfo.authChain)), deployment.origin)
```
and then delete this method

File: content/src/service/reporters/SQSDeploymentReporter.ts [hunk 3]
```diff
@@ -2,27 +2,46 @@ import log4js from "log4js";
...
 
+
+    private async onDeployment(deploymentEvent: DeploymentEvent): Promise<void> {
+        this.reportDeployment(deploymentEvent.entity,
+            Authenticator.ownerAddress(deploymentEvent.auditInfo.authChain),
+            deploymentEvent.origin)
+    }
+
+
     reportDeployment(entity: Entity, ethAddress: EthAddress, origin: string): void {
```
Comment (nchamo) [inline content/src/service/reporters/SQSDeploymentReporter.ts:45]: Could we make this private now?

File: content/src/service/reporters/SQSDeploymentReporterFactory.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { Environment, Bean } from "../../Environment"
+import { DeploymentReporter } from "./DeploymentReporter"
+import { SQSDeploymentReporter } from "./SQSDeploymentReporter"
+import { MetaverseContentService } from '../Service';
+
+export class SQSDeploymentReporterFactory {
+
+    static create(env: Environment): DeploymentReporter {
+
+        let service: MetaverseContentService = env.getBean(Bean.SERVICE)
```
Comment (nchamo) [inline content/src/service/reporters/SQSDeploymentReporterFactory.ts:10]:
```suggestion
        const service: MetaverseContentService = env.getBean(Bean.SERVICE)
```

File: content/src/service/reporters/SegmentIoAnalyticsFactory.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { Environment, Bean } from "../../Environment"
+import { DeploymentReporter } from "./DeploymentReporter"
+import { MetaverseContentService } from '../Service';
+import { SegmentIoAnalytics } from "./SegmentIoAnalytics";
+
+export class SegmentIoAnalyticsFactory {
+
+    static create(env: Environment): DeploymentReporter {
+
+        let service: MetaverseContentService = env.getBean(Bean.SERVICE)
```
Comment (nchamo) [inline content/src/service/reporters/SegmentIoAnalyticsFactory.ts:10]:
```suggestion
        const service: MetaverseContentService = env.getBean(Bean.SERVICE)
```

File: content/src/service/reporters/SegmentIoAnalytics.ts [hunk 6]
```diff
@@ -2,16 +2,32 @@ import log4js from "log4js"
...
+    }
+
+
+    private async onDeployment(deploymentEvent: DeploymentEvent): Promise<void> {
+        this.reportDeployment(deploymentEvent.entity,
+            Authenticator.ownerAddress(deploymentEvent.auditInfo.authChain),
+            deploymentEvent.origin)
     }
 
     reportDeployment(entity: Entity, ethAddress: EthAddress, origin: string): void {
```
Comment (nchamo) [inline content/src/service/reporters/SegmentIoAnalytics.ts:33]: Could you make this private?

File: content/src/service/reporters/SegmentIoAnalytics.ts [hunk 7]
```diff
@@ -2,16 +2,32 @@ import log4js from "log4js"
...
+
+            let segmentWriteKey: string = env.getConfig(EnvironmentConfig.SEGMENT_WRITE_KEY)
+            if (segmentWriteKey) {
+                this.segmentClient = new Analytics(segmentWriteKey);
+                service.listenToDeployments((deployment) => this.onDeployment(deployment));
+            }
+    }
+
+
+    private async onDeployment(deploymentEvent: DeploymentEvent): Promise<void> {
```
Comment (nchamo) [inline content/src/service/reporters/SegmentIoAnalytics.ts:27]: As before, this method can be avoided

File: content/src/service/reporters/SegmentIoAnalytics.ts [hunk 8]
```diff
@@ -2,16 +2,32 @@ import log4js from "log4js"
...
 export class SegmentIoAnalytics implements DeploymentReporter {
 
     private static readonly LOGGER = log4js.getLogger('ContentAnalyticsWithSegment');
 
     private segmentClient: Analytics;
 
-    constructor(writeKey: string) {
-        this.segmentClient = new Analytics(writeKey);
+    constructor(
+        env: Environment,
```
Comment (nchamo) [inline content/src/service/reporters/SegmentIoAnalytics.ts:16]: Again, let's try to avoid `Environment` here

File: content/test/integration/E2ETestEnvironment.ts [hunk 9]
```diff
@@ -45,7 +45,7 @@ export class E2ETestEnvironment {
             .setConfig(EnvironmentConfig.LOG_REQUESTS, false)
             .setConfig(EnvironmentConfig.LOG_LEVEL, "debug")
             .setConfig(EnvironmentConfig.BOOTSTRAP_FROM_SCRATCH, false)
-            .registerBean(Bean.DEPLOYMENT_REPORTER, new NoOpDeploymentReporter())
+            .registerBean(Bean.SEGMENT_IO_ANALYTICS, new NoOpDeploymentReporter())
```
Comment (nchamo) [inline content/test/integration/E2ETestEnvironment.ts:48]: Why is this here? Didn't you delete `NoOpDeploymentReporter`?

File: content/test/unit/service/Service.spec.ts [hunk 10]
```diff
@@ -94,7 +94,7 @@ describe("Service", function () {
             .registerBean(Bean.FAILED_DEPLOYMENTS_MANAGER, NoOpFailedDeploymentsManager.build())
             .registerBean(Bean.POINTER_MANAGER, NoOpPointerManager.build())
             .registerBean(Bean.DEPLOYMENT_MANAGER, NoOpDeploymentManager.build())
-            .registerBean(Bean.DEPLOYMENT_REPORTER, new NoOpDeploymentReporter())
+            .registerBean(Bean.SEGMENT_IO_ANALYTICS, new NoOpDeploymentReporter())
```
Comment (nchamo) [inline content/test/unit/service/Service.spec.ts:97]: Why is this here? Didn't you delete `NoOpDeploymentReporter`?

**Review by nchamo — COMMENTED** [review]
Great work!

Just a few comments 😄

File: content/src/service/reporters/SQSDeploymentReporter.ts (hunk 3 above)
Comment (agusaldasoro) [reply content/src/service/reporters/SQSDeploymentReporter.ts:45]: Does it make sense to persist the DeploymentReporter interface?

Comment (nchamo) [reply content/src/service/reporters/SQSDeploymentReporter.ts:45]: No, kill it with 🔥 !!

**Review by nchamo — APPROVED** [review]
Great wooork!!!!!
